Convert rows from Java to Ruby class

Chrizpy opened the pull request:
Related to the issue https://github.com/twingly/ecco/issues/26
Taking a piece from https://github.com/twingly/ecco/pull/56, as mentioned in that PR, the work done there should be broken down in smaller pieces, beginning with this PR.

This PR introduces `Rubyconverter` which translates that Java object in `rows` into Ruby objects

Chrizpy committed:
Add Rubyconverter

Related to the issue https://github.com/twingly/ecco/issues/26

We have our own (J)Ruby construct called "RowEvent" which holds
Ruby objects as well as one Java object that we receive from the
"mysql-binlog-connector-java".

"rows" that we receive is different depending on which SQL event is
fired. If "UPDATE" or "DELETE" is issued, a Java construct that uses
linked list that keeps a two-dimensional array is used. If "UPDATE"
is done, then a Java construct that uses array list that keeps a
map where the key is an array and the value is also an array.
The key holds the "value" before the update and the value holds the
"value" after the update, in the map.

The specs are now updated to use real Java objects instead of doubles
to test that the functionality of Rubyconverter works as intended,
which is to make the "row_event" uniform and only keep Ruby objects.

walro commented on lib/ecco/row_converter.rb:
```diff
@@ -0,0 +1,61 @@
+module Ecco
+  module RubyConverter
+    FIRST_COLUMN = 0
+    SECOND_COLUMN = 1
+    KEY = 0
+    VALUE = 1
+    ROW = 0
```
perhaps the name can be more descriptive here, like what data is found in the "first column" and so on

Chrizpy replied:
Sure

Chrizpy replied:
Fixed here: d3821566dc513a7d40e88c805a50f45202091f19

walro commented on spec/lib/ecco/row_event_listener_spec.rb:
```diff
@@ -53,6 +53,9 @@
         context "UPDATE_ROWS" do
           let(:row_event_type) { EventType::UPDATE_ROWS }
+          let(:update_entry) { Java::JavaUtil::Map.entry([], []) }
+          let(:rows) { Java::JavaUtil::ArrayList.new(Array.new(1) { update_entry }) }
```
this looks strange, a mixture between Java and Ruby objects? `Array` being Ruby.

Chrizpy replied:
Didn't think too much about wether it was a Java or a Ruby array at this stage, but yea I agree, should probably be a Java construct throughout to reflect the real world better.

Chrizpy replied:
Fixed here: 186eecbac3b2e4631996eb515e627839a6c0e8b8

walro commented on spec/event_context.rb:
```diff
@@ -2,7 +2,7 @@
 shared_context "event" do
   let(:table_id) { 1 }
-  let(:rows) { double("List") }
+  let(:rows) { Java::JavaUtil::LinkedList.new(Array.new(1) { Array.new(2) })}
```
same comment here as in https://github.com/twingly/ecco/pull/57/files#r515003860, looks like a mixture between Java and Ruby objects which probably do not reflect real-life

Chrizpy replied:
Fixed here: 186eecbac3b2e4631996eb515e627839a6c0e8b8

walro commented on spec/lib/ecco/row_event_listener_spec.rb:
```diff
@@ -53,6 +53,9 @@
 
         context "UPDATE_ROWS" do
           let(:row_event_type) { EventType::UPDATE_ROWS }
+          let(:update_entry) { Java::JavaUtil::Map.entry([], []) }
```
I can't get `Java::JavaUtil::Map.entry([], [])` to evaluate in Pry, is it used?

See:
```
[18] pry(main)> Java::JavaUtil::Map.entry([], [])
NoMethodError: undefined method `entry' for Java::JavaUtil::Map:Module
from (pry):18:in `evaluate_ruby'
```

Chrizpy replied:
Works for me:

```shell
[1] pry(main)> Java::JavaUtil::Map.entry([], [])
=> #<Java::JavaUtil::KeyValueHolder:0x3e5d4f6b>
[2] pry(main)> 
```

Are you using pry with Jruby?

walro replied:
> Works for me:

Interesting

> Are you using pry with Jruby?

Yep

```
[4] pry(main)> RUBY_PLATFORM
=> "java"
[5] pry(main)> Java::JavaUtil::Map.entry([], [])
NoMethodError: undefined method `entry' for Java::JavaUtil::Map:Module
from (pry):5:in `evaluate_ruby'
```

Did you just start `pry`? Hmm maybe I run some other JVM version

walro replied:
```
$ java -version
java version "1.8.0_261"
Java(TM) SE Runtime Environment (build 1.8.0_261-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.261-b12, mixed mode)
```

Chrizpy replied:
> Did you just start pry? Hmm maybe I run some other JVM version

Yep

```Shell
java -version
openjdk version "14.0.1" 2020-04-14
OpenJDK Runtime Environment (build 14.0.1+14)
OpenJDK 64-Bit Server VM (build 14.0.1+14, mixed mode, sharing)
```

walro replied:
I tried with more recent JVM, and it works:

```
$ JAVA_HOME=/Users/robin/Library/Java/JavaVirtualMachines/openjdk-15.0.1/Contents/Home pry
[1] pry(main)> Java::JavaUtil::Map.entry([], [])
=> #<Java::JavaUtil::KeyValueHolder:0x5e95b0ab>
```

However, I believe we should only use things that works with Java 8 (as that's the oldest supported version of JVM for JRuby). We should likely change our travis config to run with version 8 too.

walro replied:
Moreover, the real implementation uses an `AbstractMap.SimpleEntry`, see: 

https://github.com/osheroff/mysql-binlog-connector-java/blob/master/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/UpdateRowsEventDataDeserializer.java#L70

It just just cast to a `Map.Entry`.

Chrizpy replied:
Changed to correct entry here: 186eecbac3b2e4631996eb515e627839a6c0e8b8

walro commented on lib/ecco/row_converter.rb:
```diff
@@ -0,0 +1,61 @@
+      end
+    end
+
+    def self.java_linked_list_to_ruby_array(java_rows)
```
I think that things potentially can be simplified:

```
[1] pry(main)> java_list = Java::JavaUtil::LinkedList.new
=> #<Java::JavaUtil::LinkedList:0x4372b9b6>
[2] pry(main)> ruby_list = java_list.to_a
=> []
[3] pry(main)> ruby_list.class
=> Array
[4] pry(main)> java_list.class
=> Java::JavaUtil::LinkedList
```

I haven't tested more than the above though.

Chrizpy replied:
Interesting, I will look more into this as well.

Chrizpy replied:
Changed here: d3821566dc513a7d40e88c805a50f45202091f19

Chrizpy commented on lib/ecco/row_converter.rb:
```diff
@@ -0,0 +1,61 @@
+module Ecco
+  module RubyConverter
```
I just noticed now that the file is `row_converter` but the module is called `RubyConverter`. I should probably fix this so that it is cohesive somehow.

walro replied:
Yeah, `RowsConverter` would make sense to me at least. (so file would be `rows_converter.rb`)

Chrizpy replied:
Fixed here: 5f2d1e5735964f985c46fd85b71527fda9a5cca8

Chrizpy committed:
Fix name of new module

Chrizpy committed:
More correct mock datatypes

Chrizpy committed:
Simplify converter and better naming

Chrizpy committed:
Change method name

To something more describing instead of being technical

Chrizpy committed:
Convert whole array instead of deep diving

Chrizpy committed:
Correct mocks

Chrizpy commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Sprint87: Abstrahera bort Java-objekt i rader](https://trello.com/c/EBXf8FPN/3353-sprint87-abstrahera-bort-java-objekt-i-rader)

Chrizpy committed:
Fix require pattern

roback commented on spec/lib/ecco/row_event_listener_spec.rb:
```diff
@@ -53,6 +56,17 @@
 
         context "UPDATE_ROWS" do
           let(:row_event_type) { EventType::UPDATE_ROWS }
+          let(:rows) do
+            ArrayList.new(
+              Arrays.asList(
+                AbstractMap::SimpleEntry.new(
+                  Array.new(2).to_java,
+                  Array.new(2).to_java
+                )
+              )
+            )
+          end
+
```
I tried the following to get rid of the `let(:rows)` business from this file. Just to make sure all rows are defined in the same file (`event_context.rb`). The imports needs to move too.

WDYT?

This could perhaps be cleaned up a bit, but I have no good ideas right now.

```diff
diff --git a/spec/event_context.rb b/spec/event_context.rb
index d24c178..a303d81 100644
--- a/spec/event_context.rb
+++ b/spec/event_context.rb
@@ -3,13 +3,36 @@ java_import java.util.Arrays
 
 shared_context "event" do
   let(:table_id) { 1 }
-  let(:rows) do
+
+  def update_event?
+    [
+      EventType::EXT_UPDATE_ROWS,
+      EventType::UPDATE_ROWS,
+    ].include?(row_event_type)
+  end
+
+  let(:rows_for_update_event) do
+    ArrayList.new(
+      Arrays.asList(
+        AbstractMap::SimpleEntry.new(
+          Array.new(2).to_java,
+          Array.new(2).to_java
+        )
+      )
+    )
+  end
+
+  let(:rows_for_write_and_delete_events) do
     Java::JavaUtil::LinkedList.new(
       Arrays.asList(
         Array.new(1).to_java { Array.new(2).to_java }
       )
     )
   end
+
+  let(:rows) do
+    update_event? ? rows_for_update_event : rows_for_write_and_delete_events
+  end
   let(:database) { "some_database" }
   let(:table) { "some_table" }
   let(:table_event_type) { EventType::TABLE_MAP }
diff --git a/spec/lib/ecco/row_event_listener_spec.rb b/spec/lib/ecco/row_event_listener_spec.rb
index a476845..e051b2c 100644
--- a/spec/lib/ecco/row_event_listener_spec.rb
+++ b/spec/lib/ecco/row_event_listener_spec.rb
@@ -56,17 +56,6 @@ describe Ecco::RowEventListener do
 
         context "UPDATE_ROWS" do
           let(:row_event_type) { EventType::UPDATE_ROWS }
-          let(:rows) do
-            ArrayList.new(
-              Arrays.asList(
-                AbstractMap::SimpleEntry.new(
-                  Array.new(2).to_java,
-                  Array.new(2).to_java
-                )
-              )
-            )
-          end
-
           it "should return UPDATE_ROWS" do
             actual_type = nil
             subject.callback = Proc.new do |row_event|
@@ -111,17 +100,6 @@ describe Ecco::RowEventListener do
 
         context "EXT_UPDATE_ROWS" do
           let(:row_event_type) { EventType::EXT_UPDATE_ROWS }
-          let(:rows) do
-            ArrayList.new(
-              Arrays.asList(
-                AbstractMap::SimpleEntry.new(
-                  Array.new(2).to_java,
-                  Array.new(2).to_java
-                )
-              )
-            )
-          end
-
           it "should return UPDATE_ROWS" do
             actual_type = nil
             subject.callback = Proc.new do |row_event|
```

Chrizpy replied:
Oh nice! This is very helpful, gonna take a closer look soon. :)

Chrizpy replied:
I don't know how to do this cleaner either, gonna check a little more. Overall I think this change is good, makes the spec file a bit less cluttered.

Chrizpy replied:
```diff
diff --git a/spec/event_context.rb b/spec/event_context.rb
index d24c178..e0623e7 100644
--- a/spec/event_context.rb
+++ b/spec/event_context.rb
@@ -3,17 +3,39 @@ java_import java.util.Arrays
 
 shared_context "event" do
   let(:table_id) { 1 }
+  let(:database) { "some_database" }
+  let(:table) { "some_table" }
+  let(:table_event_type) { EventType::TABLE_MAP }
+  let(:row_event_type) { EventType::EXT_WRITE_ROWS }
   let(:rows) do
+    update_event? ? rows_for_update_event : rows_for_write_and_delete_events
+  end
+
+  def update_event?
+    [
+      EventType::EXT_UPDATE_ROWS,
+      EventType::UPDATE_ROWS,
+    ].include?(row_event_type)
+  end
+
+  let(:rows_for_update_event) do
+    ArrayList.new(
+      Arrays.asList(
+        AbstractMap::SimpleEntry.new(
+          Array.new(2).to_java,
+          Array.new(2).to_java
+        )
+      )
+    )
+  end
+
+  let(:rows_for_write_and_delete_events) do
     Java::JavaUtil::LinkedList.new(
       Arrays.asList(
         Array.new(1).to_java { Array.new(2).to_java }
       )
     )
   end
-  let(:database) { "some_database" }
-  let(:table) { "some_table" }
-  let(:table_event_type) { EventType::TABLE_MAP }
-  let(:row_event_type) { EventType::EXT_WRITE_ROWS }
 
   let(:table_event) do
     event = double("Event")

```

Chrizpy replied:
I rearranged the order a little bit, maybe this is better? :)

roback replied:
:+1: Yes, I think that looks better having the `let`s grouped that way :)

Chrizpy replied:
Bam 436955b6e3ee90c599ec90c9176f2d6d82b05c0e

Chrizpy committed:
Clean up row_event_listener_spec.rb

Move the different row definitions into the event_context.rb instead.

Co-Authored-By: Mattias Roback <2803721+roback@users.noreply.github.com>

roback commented on spec/lib/ecco/row_event_listener_spec.rb:
```diff
@@ -1,3 +1,6 @@
+java_import java.util.ArrayList
+java_import java.util.AbstractMap
+
```
These can be moved to `event_context.rb` (Did not do that in https://github.com/twingly/ecco/pull/57#discussion_r524397737 because lazy :))

Chrizpy replied:
👍  00b9350

roback commented on lib/ecco/rows_converter.rb:
```diff
@@ -0,0 +1,50 @@
+    end
+  end
+
+  class Tuple
```
Maby we can come up with a better name for this class. What it contains are the column values before and after.

Can't say I have any good ideas right now though 🤔

roback replied:
Just some ideas regarding this class. The name is still pending, but based on the findings in https://github.com/twingly/ecco/pull/57#discussion_r529630356 we should be able to come up with something descriptive.

```diff
diff --git a/lib/ecco/rows_converter.rb b/lib/ecco/rows_converter.rb
index 3b4582e..1f5b611 100644
--- a/lib/ecco/rows_converter.rb
+++ b/lib/ecco/rows_converter.rb
@@ -13,38 +13,29 @@ module Ecco
     end

     def self.convert_update_rows(java_rows)
-      before_update_row = java_rows.first.get_key.to_a
-      after_update_row  = java_rows.first.get_value.to_a
+      java_rows.map do |java_row|
+        before_update_row = java_row.get_key.to_a
+        after_update_row  = java_row.get_value.to_a

-      Array.new(1) do
-        Tuple.new(
-          key: before_update_row,
-          value: after_update_row
+        UpdateRowsEventUpdatedRow.new(
+          before: before_update_row,
+          after:  after_update_row
         )
       end
     end
   end

-  class Tuple
-    KEY   = 0
-    VALUE = 1
+  class UpdateRowsEventUpdatedRow
+    attr_reader :before
+    attr_reader :after

-    attr_reader :tuple
-
-    def initialize(key:, value:)
-      @tuple = [key, value]
-    end
-
-    def key
-      @tuple[KEY]
+    def initialize(before:, after:)
+      @before = before
+      @after  = after
     end

-    def value
-      @tuple[VALUE]
-    end
-
-    def inspect
-      @tuple.to_s
-    end
+    # Make sure we don't break existing code
+    alias_method :key, :before
+    alias_method :value, :after
   end
 end
```

roback replied:
Note: I removed the inspect above, since I thought the default was good enough, but we could define our own if needed. What the default looks like:

```
#<Ecco::UpdateEventColumnsDiff:0x3667faa8 @after=[1, "another value"], @before=[1, "a value"]>
```

Chrizpy replied:
I really like the idea of using aliases to make our code more readable from the inside and still work with the code from the binlog java stuff. Wish I had thought of that! :)

Chrizpy replied:
Added your smart idea here: e0f1331

roback commented on lib/ecco/rows_converter.rb:
```diff
@@ -0,0 +1,50 @@
+      java_rows.to_a
+    end
+
+    def self.convert_update_rows(java_rows)
+      before_update_row = java_rows.first.get_key.to_a
+      after_update_row  = java_rows.first.get_value.to_a
+
+      Array.new(1) do
+        Tuple.new(
+          key: before_update_row,
+          value: after_update_row
+        )
+      end
+    end
```
What `java_rows` above contains is an array containing each row that was updated. Each of those rows has the value of all columns before and after the update. Because of this we can't just call `java_rows.first`, we need to loop through every row in `java_rows` and create a `Tuple` for each of them. That is, something along the lines of:

```diff
diff --git a/lib/ecco/rows_converter.rb b/lib/ecco/rows_converter.rb
index 3b4582e..7647ff2 100644
--- a/lib/ecco/rows_converter.rb
+++ b/lib/ecco/rows_converter.rb
@@ -13,10 +13,10 @@ module Ecco
     end

     def self.convert_update_rows(java_rows)
-      before_update_row = java_rows.first.get_key.to_a
-      after_update_row  = java_rows.first.get_value.to_a
+      java_rows.map do |java_row|
+        before_update_row = java_row.get_key.to_a
+        after_update_row  = java_row.get_value.to_a

-      Array.new(1) do
         Tuple.new(
           key: before_update_row,
           value: after_update_row
```

But before attempting something like that we would need a test which updates multiple rows. All tests we have now only updates one row at a time, which is why it worked before :)

Chrizpy replied:
We are taking all indexes into consideration here: e0f1331

walro commented:
I also mentioned this somewhere before (IRL?) but I think some kind of "deep inspect spec" on the resulting objects would be good. Just to ensure that we have found all Java traces. Ensure that all different type of "rows" are checked as I think it differs between INSERT/UPDATE/DELETE/YOUNAMEIT

Also (perhaps another PR building on this one), everything is basically mapped 1:1 with the Java implementation. I feel that the Java version is a bit obsessive with using primitive/built-in types rather than introducing better data structures (why we end up with awkward things as lists with "pairs" with lists)

Chrizpy committed:
Move imports to event_context.rb

Co-Authored-By: Mattias Roback <2803721+roback@users.noreply.github.com>

Chrizpy committed:
Refactor from PR feedback

UpdateRowsEventUpdatedRow was maybe not the
greatest name, but lack for a better one I'll let it
stay for now.

Co-Authored-By: Mattias Roback <2803721+roback@users.noreply.github.com>

walro commented:
I hope I got the merge right in 94a8df6, the specs passed locally at least :)

Chrizpy committed:
Add specs

Comment: https://github.com/twingly/ecco/pull/57#issuecomment-733060581

Chrizpy commented:
Heh, sorry for reviving this. :P
